Support multiple selections in search fields

inguin opened the pull request:
* Convert the search fields in the quick filter toolbar and advanced
  search to multi-select fields
* For saved searches, store the selection as string of comma-separated
  values
* Add DB migration to change column types in `custom_filter` to string
* List of changed fields:
  - Assigned (cst_users)
  - Category (cst_iss_prc_id)
  - Priority (cst_iss_pri_id)
  - Product  (cst_pro_id)
  - Release  (cst_iss_pre_id)
  - Reporter (cst_reporter)
  - Severity (cst_iss_sev_id)
  - Status   (cst_iss_sta_id)

glensc commented:
don't like the idea of having `_id` column of type string. @balsdorf what you think?

inguin commented:
I agree with @glensc, a column named 'cst_iss_sta_id' might be expected to directly reference 'iss_sta_id' in the issue table.

I tried to keep the schema changes to a minimum, but it should't be a big deal to rename the columns if you'd prefer that.

balsdorf commented:
I'll do a deeper dive into the code later but like @glensc I agree that the column names should be changed.   I'd be okay with cst_categories, cst_severities, etc

inguin commented:
I renamed the `custom_filter` columns according to your recommendations. I also split the feature into two commits, one for the DB migration and one for the actual multi-select search.

glensc commented on db/migrations/20171109145542_eventum_custom_filter_multiselect.php:
```diff
@@ -0,0 +1,47 @@
+        'cst_iss_sev_id' => 'cst_severities',
+        'cst_reporter'   => 'cst_reporters',
+        'cst_iss_prc_id' => 'cst_categories',
+        'cst_iss_sta_id' => 'cst_status',
```
should be `cst_statuses` then if it contains multiple values like other fields.

inguin replied:
I always have trouble making up my mind about that word. The grammatically correct plural is 'status', but it sounds weird and might conflict with a singular use. I'll change it.

glensc commented on db/migrations/20171109145542_eventum_custom_filter_multiselect.php:
```diff
@@ -0,0 +1,47 @@
+    {
+        $table = $this->table('custom_filter');
+        foreach (self::COLUMNS as $old_name => $new_name) {
+            $table->changeColumn($old_name, 'string', ['length' => 64, 'null' => true]);
```
i would use `255` here as there's no special reason to have `64`

inguin replied:
ACK

inguin commented:
I renamed  `cst_status` to `cst_statuses` and changed the length of string fields from 64 to 255. I also added a migration step for `cst_users` which for some reason already was a varchar(64) before.

glensc commented:
tried the branch locally (4ea804435), but get dozen of warning notices.

db patch applied
```
 == 20171109145542 EventumCustomFilterMultiselect: migrating
```

```
( ! ) Warning: Invalid argument supplied for foreach() in /home/glen/scm/eventum/eventum/lib/eventum/class.search.php on line 502
--


1 | 0.0003 | 253800 | {main}( ) | .../list.php:0
2 | 0.0777 | 5674544 | Eventum\Controller\BaseController->run( ) | .../list.php:17
3 | 0.0949 | 6969528 | Eventum\Controller\ListController->prepareTemplate( ) | .../BaseController.php:78
4 | 0.1072 | 7210520 | Search::getListing( ) | .../ListController.php:206
5 | 0.1153 | 7868960 | Search::buildWhereClause( ) | .../class.search.php:336


```

inguin commented:
Added a commit to fix all warnings. My modified `Search::getParam()` wasn't very good at assuring the return values would be arrays; default values or values from the user's search profile could still be returned as scalars.

I'll squash the fixes into the actual implementation commit once the pull request is approved.

glensc commented:
checked out the branch locally.

i don't like that the ui change makes search filters taking 50% of the screen.

such large options block belongs to advanced search imho. could the default view be still single field based but allow multiselects on advanced search? and advocate using saved searches for common filters?


before:
![image](https://user-images.githubusercontent.com/199095/32983808-82918b66-cca3-11e7-8023-b559f6f6cbb4.png)

after:
![image](https://user-images.githubusercontent.com/199095/32983793-44af44aa-cca3-11e7-9ab3-83f1d41f840a.png)

inguin commented:
I understand your concern (although you'd have to have a pretty low screen resolution for the search bar to fill half of the height). We've been using the tall filter fields without complains for at least a decade, but that doesn't mean other long-time Eventum users will be happy with the change. I could reduce the number of rows in the select boxes, but that makes then less convenient to use.

Limiting the multi-selection to the Advanced Search page makes sense and would be very easy to implement. The only issue is that the user won't be able to repeat or refine the search afterwards. Clicking the Search button will lose part of his selection.

A possible solution for this would be to introduce a special 'multiple' option that is displayed and selected for search fields with more than one option selected.

inguin commented:
I added an experimental commit reverting the quick filter back back to plain select boxes and introducing a special "multiple selections" item.

glensc commented:
looks good now, i think it's ok to be merged
